### DISCUSSION TITLE
Using Depends() in other functions, outside the endpoint path operation!

### DISCUSSION THREAD

**Discussion by vjanz** [fastapi/fastapi#7720] [Questions]
Hey community,

I don't know if this feature is available, but I want to ask if I can use dependencies outside endpoint path decorators and functions..

I have this repository, when inside some function on that, I need another repository function to perform the full task.
What I want to do is to inject that as a dependency, and not to create an instance of it, since that is best way to do that here in python, because I come from background of Java and I always used to initialize  a instance of that Class.

So basically..I have a function to get an instance of SynonymCrud()
```
def get_synonym_crud():
    return synonym_crud
```

the function in to create a word, in #word_crud.py
```
def create_word(self, word: WordCreate, dictionary_id: int, synonym: List[SynonymCreate],synonym_crud:SynonymCrud = Depends(get_synonym_crud)):
w_dict = word.dict()
w_dict.update({"word": word.word.lower()})
db_word = Word(**w_dict, dictionary_id=dictionary_id)
self.database.add(db_word)
self.database.commit()
self.database.refresh(db_word)
synonyms_to_create = synonym_crud.create_many_synonyms(synonym, db_word.id)
return db_word
```
Here I tried to inject: synonym_crud:SynonymCrud = Depends(get_synonym_crud), 
but when I try to execute this on the endpoint I get an error, that says:
`
AttributeError: 'Depends' object has no attribute 'create_many_synonyms',` even tho I got this function in the SynonymRepository.

This is not the only case, even when I try to inject anything else like database dependency outside the endpoint I get like: `AttributeError: 'Depends' object has no attribute 'query' `or something similar.

endpoint..
```
@router.post("/tests/words/", response_model=Word)
def create_word(request: Request, word: WordCreate, dictionary_id: int,
synonym: List[SynonymCreate],
logger: logging.Logger = Depends(get_logger_test),
word_repo: WordCrud = Depends(get_word_crud),
current_user: User = Depends(get_curr_active_superuser)):
word_to_create = word_repo.find_by_name(word.word)
if word_to_create:
raise HTTPException(status_code=400, detail="Word already exists")
word_created = word_repo.create_word(word, dictionary_id, synonym)

    return word_created
```
Here the other dependency word_repo, works without any error, but when I try to inject anything outside the endpoint functions( like the example with synonym).

Is this possible, or what am I doing wrong?
Thank you in advance

**Comment by tiangolo** [comment]
Hey @vjanz , if your code runs outside of FastAPI you're better using explicit objects.

Dependencies are, in the end, a way to get or process data based on a request. But if your code is not related to a request (not in a FastAPI app) then there's no advantage and it would just add complexity to your code.

On the other side, a dependency function is in the end, just a function. So, if you pass all the arguments with the correct values instead of letting it get the default value returned by `Depends()` you can just use it directly as a normal function.

Anyway, I assume you solved your problem, so, thanks for closing the issue :+1:

**Comment by rfleschenberg** [comment]
One situation where this would be useful is if you have code paths outside of router functions (the equivalent of Django management commands, basically) that you are calling from tests, and you want to share your `dependency_overrides` with them.

I guess I can use monkey-patching / `unittest.mock` instead, but having `dependency_overrides` "just work" would be nice.

**Comment by m4bwav** [comment]
A good dependency injection system can have value outside of a web request.

**Comment by unkindypie** [comment]
> A good dependency injection system can have value outside of a web request.

Totally agree. I have a dozen services built on top of FastAPI's DI system and I simply can't use them without manually setting up all of their dependencies if I would like to run a service from a CRON task which isn't a part of FastAPI app.

**Comment by couteau** [comment]
Agree that when you have lots of dependencies and sub dependencies, setting them up manually is not ideal. I came up with a quick and dirty solution to this that works for my use case (basically, using the same set of services for command line scripts as for api endpoints). Not sure if it will work for all use cases -- e.g., I'm not sure what would happen if any dependencies depended on query parameters, path parameters, headers, or cookies.

It looks like this:

```
import argparse
import inspect
from abc import abstractmethod
from contextlib import AsyncExitStack
from typing import Callable

from fastapi import Request
from fastapi.dependencies.models import Dependant
from fastapi.dependencies.utils import (
    get_dependant,
    solve_dependencies
)
from pydantic import ValidationError


class Command:
    def __init__(self, name: str | None = None, namespace=None):
        self.name = name or self.__class__.__name__
        self.parser = self.make_parser()
        self.args = self.parser.parse_args(namespace=namespace)

    # override to add command-line arguments
    def make_parser(self) -> argparse.ArgumentParser:
        return argparse.ArgumentParser(prog=self.name)

    @abstractmethod
    async def command(self):
        pass

    async def solve_command(self, request: Request, dependant: Dependant):
        values, errors, _1, _2, _3 = await solve_dependencies(request=request, dependant=dependant)
        if errors:
            raise ValidationError(errors, None)
        if inspect.iscoroutinefunction(dependant.call):
            result = await dependant.call(**values)
        else:
            result = dependant.call(**values)

        return result

    async def run_command(self, command: Callable):
        async with AsyncExitStack() as cm:
            request = Request({
                "type": "http",
                "headers": [],
                "query_string": "",
                "fastapi_astack": cm
            })

            dependant = get_dependant(
                path=f"command:{self.name}",
                call=command
            )

            return await self.solve_command(request, dependant)

    async def run(self):
        return await self.run_command(self.command)
```
A subclass can override the `command` method and add/use any top-level dependencies. The dependencies will then be resolved and the `command` method executed when the script calls the `run` method. Or an arbitrary function with dependencies can be invoked using the `run_command` method.

E.g.,
```
import asyncio
from fastapi import Depends
from app.services import MyService
from app.command import Command

class MyCommand(Command):
    async def command(self, service: MyService = Depends()):
        service.DoSomething()

def main():
    command = MyCommand("my-command")
    asyncio.run(command.run())

if __name__ == "__main__":
    main()
```
or
```
import asyncio
from fastapi import Depends
from app.services import MyService
from app.command import Command

async def command(service: MyService = Depends()):
    service.DoSomething()

def main():
    command = Command("my-command")
    asyncio.run(command.run_command(command))

if __name__ == "__main__":
    main()
```

**Comment by fabiocerqueira** [comment]
Is there any plan to be able to wire other parts of the code with the FastAPI DI?

**Comment by agn-7** [comment]
Can anybody tell me how to use a dependency injection for my `get_db()` outside of the FastAPI routers? 

Here's the get_db async generator:

```python
async def get_db() -> AsyncGenerator[AsyncSession, None]:
    async with async_session() as session:
        yield session
```

In the FastAPI router I can simply use `Depends()` like this:

```python
@router.get("/interactions", response_model=List[schemas.Interaction])
async def get_all_interactions(db: Annotated[AsyncSession, Depends(get_db)]) -> List[schemas.Interaction]:
    interactions = await crud.get_interactions(db=db)

    return [
        schemas.Interaction.model_validate(interaction) for interaction in interactions
    ]
```

Now, outside of the request how can I inject the `get_db` within the method and get rid of `async for`?

```python
@cli.command(name="create_superuser")
async def create_superuser():
    username = click.prompt("Username", type=str)
    email = click.prompt("Email (optional)", type=str, default="")
    password = getpass("Password: ")
    confirm_password = getpass("Confirm Password: ")

    if password != confirm_password:
        click.echo("Passwords do not match")
        return

    async for db in database.get_db():  # Note
        user = schemas.UserAdminCreate(
            username=username,
            email=None if not email else email,
            password=password,
            role="admin",
        )
        await crud.create_user(db=db, user=user)
```

**Reply by agn-7** [reply]
Here I proposed a dependency injector for resolving the `get_db()` which is an async generator:

https://stackoverflow.com/a/77821629/3702377

**Reply by jamie-at-arkifi** [reply]
"invent my own dependency injector" isn't the solution I'm hoping for.

But honestly, I might. If only so I can implement interop with pytest fixtures.

**Reply by agn-7** [reply]
Here I also proposed a full workaround with pytest and my dependency injector:

https://stackoverflow.com/a/77858844/3702377


but generally, you can simply define a db session fixture on your test side and pass it to the method having the dependency injector to avoid the default value and get the passing db session.

**Comment by jamie-at-arkifi** [comment]
Yup, I can think of several reasons you'd want dependencies outside of a web request:

* CLI-based management commands
* Job queue execution
* Development tooling
* Testing

**Comment by spermiakov** [comment]
I have one more use-case. We have RequestContext wrapper as request scoped object, this object has some request scoped tags which we include in every custom metric we submit to cloudwatch/datadog/elk. In current scenario we have to pass this object around all method/functions as it not possible to inject it freely for functions which is outside of entry point @app.get/post etc. All workaround for DI also wont work at it work more as proto/singleton and not request scope. As fastapi is NIO threadlocal also no an option, I'm not even sure if this concept exist in python(it's from java world). As workaround you can maintain some dict with uuid generated request-id to request, but amount of effort to maintain this structure will neglect all benefits of using it.
Another use-case is user object you don't want to pass all around which is also RequestScoped(for some is session but this is more complex scenario)

**Reply by agn-7** [reply]
But just in case, there are PyPI packages namely [dependency-injector](https://github.com/ets-labs/python-dependency-injector), [injector](https://github.com/python-injector/injector), and [pinject](https://github.com/google/pinject) in this regard.

**Reply by spermiakov** [reply]
This project provides only Singleton and prototype scopes, I don't know how that project will wire RequestScope, as it lifecycle in FastApi control.

**Reply by tekumara** [reply]
https://github.com/hynek/svcs has containers that provide services for the life of a request

**Comment by barapa** [comment]
In case this is useful, here is the current method I am using:

```python
import inspect
from contextlib import AsyncExitStack
from functools import wraps
from typing import Any, Callable, Coroutine, TypeVar

from fastapi import Request
from fastapi.dependencies.models import Dependant
from fastapi.dependencies.utils import get_dependant, solve_dependencies
from loguru import logger

T = TypeVar("T")


class DependencyError(Exception):
    """Exception raised for errors during dependency injection."""

    pass


def injectable(
    func: Callable[..., Coroutine[Any, Any, T]],
) -> Callable[..., Coroutine[Any, Any, T]]:
    """
    A decorator to enable FastAPI-style dependency injection for any asynchronous function.
    This allows dependencies defined with FastAPI's Depends mechanism to be automatically
    resolved and injected into CLI tools or other components, not just web endpoints.
    Args:
        func: The asynchronous function to be wrapped, enabling dependency injection.
    Returns:
        The wrapped function with dependencies injected.
    Raises:
        ValueError: If the dependant.call is not a callable function.
        RuntimeError: If the wrapped function is not asynchronous.
        DependencyError: If an error occurs during dependency resolution.
    """

    @wraps(func)
    async def call_with_solved_dependencies(*args: Any, **kwargs: Any) -> T:  # type: ignore
        dependant: Dependant = get_dependant(path="command", call=func)
        if dependant.call is None or not callable(dependant.call):
            raise ValueError("The dependant.call attribute must be a callable.")

        if not inspect.iscoroutinefunction(dependant.call):
            raise RuntimeError("The decorated function must be asynchronous.")

        fake_request = Request({"type": "http", "headers": [], "query_string": ""})
        values: dict[str, Any] = {}
        errors: list[Any] = []

        async with AsyncExitStack() as stack:
            solved_result = await solve_dependencies(
                request=fake_request,
                dependant=dependant,
                async_exit_stack=stack,
            )
            values, errors = solved_result[0], solved_result[1]

            if errors:
                error_details = "\n".join([str(error) for error in errors])
                logger.info(f"Dependency resolution errors: {error_details}")

            return await dependant.call(*args, **{**values, **kwargs})

    return call_with_solved_dependencies
```

And here is an example of using it in a CLI built with the [cyclopts](https://cyclopts.readthedocs.io/en/latest/) library.
```python
from cyclopts import App
from fastapi import Depends
from loguru import logger

from app.clis.lib.injectable import injectable
from app.deps.settings import Settings, provide_settings

app = App()

@app.default
@injectable
async def example(
    *,
    message: str,
    settings: Settings = Depends(provide_settings),
) -> None:
    """Example command using injectable with cyclopts"""
    logger.info(message)
    logger.warning(settings.generate_db_url())


if __name__ == "__main__":
    app()
```

**Reply by spermiakov** [reply]
I briefly checked solve_dependencies and it seems like it does not replace your fake_request with valid one, I want our services to be as stateless as possible and use stateful information only when it needed on demand. Singletons and prototypes are easy to implement with factory class, for smaller projects it suit better.

**Reply by blalor** [reply]
This is indeed super useful. Thank you @barapa!! I'm planning to use it for RQ-invoked tasks and it seems to be working perfectly. I'm experimenting with caching dependencies, as that's about the only thing I see missing, but if that's not important this does the job.

**Reply by barapa** [reply]
If you make any improvements, please post them.

**Reply by spermiakov** [reply]
I ended up using 

```
from contextvars import ContextVar

request_context_holder: ContextVar[RequestContext] = ContextVar(
    "rq"
)

class RequestContext:
        def __init__(self):
                self.__context_var_token = None

        async def __aenter__(self):
        self.__context_var_token = request_context_holder.set(self)
        return self

    async def __aexit__(self, *err):
        if self.__context_var_token:
            request_context_holder.reset(self.__context_var_token)
            self.__context_var_token = None
---------
      Middleware 
                    async def dispatch(self, request: Request, call_next):
                             with new RequestContext():
                                       return await call_next(request)
```
injector is using contextvars under the hood for request scoped and ths natively supported by python for asyncio

**Comment by notypecheck** [comment]
It would be better to just use a separate DI framework/library, there's a lot of theme out there: https://github.com/sfermigier/awesome-dependency-injection-in-python?tab=readme-ov-file#di-frameworks--containers

**Comment by JasperSui** [comment]
> In case this is useful, here is the current method I am using:
> 
> ```python
> import inspect
> from contextlib import AsyncExitStack
> from functools import wraps
> from typing import Any, Callable, Coroutine, TypeVar
> 
> from fastapi import Request
> from fastapi.dependencies.models import Dependant
> from fastapi.dependencies.utils import get_dependant, solve_dependencies
> from loguru import logger
> 
> T = TypeVar("T")
> 
> 
> class DependencyError(Exception):
>     """Exception raised for errors during dependency injection."""
> 
>     pass
> 
> 
> def injectable(
>     func: Callable[..., Coroutine[Any, Any, T]],
> ) -> Callable[..., Coroutine[Any, Any, T]]:
>     """
>     A decorator to enable FastAPI-style dependency injection for any asynchronous function.
>     This allows dependencies defined with FastAPI's Depends mechanism to be automatically
>     resolved and injected into CLI tools or other components, not just web endpoints.
>     Args:
>         func: The asynchronous function to be wrapped, enabling dependency injection.
>     Returns:
>         The wrapped function with dependencies injected.
>     Raises:
>         ValueError: If the dependant.call is not a callable function.
>         RuntimeError: If the wrapped function is not asynchronous.
>         DependencyError: If an error occurs during dependency resolution.
>     """
> 
>     @wraps(func)
>     async def call_with_solved_dependencies(*args: Any, **kwargs: Any) -> T:  # type: ignore
>         dependant: Dependant = get_dependant(path="command", call=func)
>         if dependant.call is None or not callable(dependant.call):
>             raise ValueError("The dependant.call attribute must be a callable.")
> 
>         if not inspect.iscoroutinefunction(dependant.call):
>             raise RuntimeError("The decorated function must be asynchronous.")
> 
>         fake_request = Request({"type": "http", "headers": [], "query_string": ""})
>         values: dict[str, Any] = {}
>         errors: list[Any] = []
> 
>         async with AsyncExitStack() as stack:
>             solved_result = await solve_dependencies(
>                 request=fake_request,
>                 dependant=dependant,
>                 async_exit_stack=stack,
>             )
>             values, errors = solved_result[0], solved_result[1]
> 
>             if errors:
>                 error_details = "\n".join([str(error) for error in errors])
>                 logger.info(f"Dependency resolution errors: {error_details}")
> 
>             return await dependant.call(*args, **{**values, **kwargs})
> 
>     return call_with_solved_dependencies
> ```
> 
> And here is an example of using it in a CLI built with the [cyclopts](https://cyclopts.readthedocs.io/en/latest/) library.
> 
> ```python
> from cyclopts import App
> from fastapi import Depends
> from loguru import logger
> 
> from app.clis.lib.injectable import injectable
> from app.deps.settings import Settings, provide_settings
> 
> app = App()
> 
> @app.default
> @injectable
> async def example(
>     *,
>     message: str,
>     settings: Settings = Depends(provide_settings),
> ) -> None:
>     """Example command using injectable with cyclopts"""
>     logger.info(message)
>     logger.warning(settings.generate_db_url())
> 
> 
> if __name__ == "__main__":
>     app()
> ```

Based on the version and creative idea provided by @barapa earlier, I’ve implemented a more complete version with caching considerations, and also support both async and sync functions.

The version below is compatible as of `python=3.12.4` and `fastapi[standard]=0.115.0`.

Now, we can use `use_cache` to decide whether or not to reuse already instantiated sub-dependencies (similar to the original `use_cache` mechanism in FastAPI’s Depends, but please correct me if I’ve misunderstood anything).

Here’s the full implementation:

```python
import asyncio
import inspect
import logging
from collections.abc import Callable, Coroutine
from contextlib import AsyncExitStack
from functools import wraps
from typing import Any, TypeVar, cast

from fastapi import Request
from fastapi.dependencies.models import Dependant
from fastapi.dependencies.utils import get_dependant, solve_dependencies

logger = logging.getLogger(__name__)
T = TypeVar("T")
_SOLVED_DEPENDENCIES: dict[tuple[Callable[..., Any], tuple[str]], Any] = {}


class DependencyError(Exception):
    """Exception raised for errors during dependency injection."""

def injectable(  # noqa: C901
    func: Callable[..., T] | Callable[..., Coroutine[Any, Any, T]] | None = None,
    *,
    use_cache: bool = True,
) -> Callable[..., T] | Callable[..., Coroutine[Any, Any, T]]:
    """A decorator to enable FastAPI-style dependency injection for any function (sync or async).

    This allows dependencies defined with FastAPI's Depends mechanism to be automatically
    resolved and injected into CLI tools or other components, not just web endpoints.

    Args:
        func: The function to be wrapped, enabling dependency injection.
        use_cache: Whether to use the dependency cache for the arguments a.k.a sub-dependencies.

    Returns:
        The wrapped function with dependencies injected.

    Raises:
        ValueError: If the dependant.call is not a callable function.
        DependencyError: If an error occurs during dependency resolution.
    """

    def _impl(
        func: Callable[..., T] | Callable[..., Coroutine[Any, Any, T]],
    ) -> Callable[..., T] | Callable[..., Coroutine[Any, Any, T]]:
        is_async = inspect.iscoroutinefunction(func)
        dependency_cache = _SOLVED_DEPENDENCIES if use_cache else None

        async def resolve_dependencies(dependant: Dependant) -> tuple[dict[str, Any], list[Any] | None]:
            fake_request = Request({"type": "http", "headers": [], "query_string": ""})
            async with AsyncExitStack() as stack:
                solved_result = await solve_dependencies(
                    request=fake_request,
                    dependant=dependant,
                    async_exit_stack=stack,
                    embed_body_fields=False,
                    dependency_cache=dependency_cache,
                )
                dep_kwargs = solved_result.values  # noqa: PD011
                if dependency_cache is not None:
                    dependency_cache.update(solved_result.dependency_cache)

            return dep_kwargs, solved_result.errors

        def handle_errors(errors: list[Any] | None) -> None:
            if errors:
                error_details = "\n".join(str(error) for error in errors)
                logger.info(f"Dependency resolution errors: {error_details}")

        def validate_dependant(dependant: Dependant) -> None:
            if dependant.call is None or not callable(dependant.call):
                msg = "The dependant.call attribute must be a callable."
                raise ValueError(msg)

        @wraps(func)
        async def async_call_with_solved_dependencies(*args: Any, **kwargs: Any) -> T:  # noqa: ANN401
            dependant = get_dependant(path="command", call=func)
            validate_dependant(dependant)
            deps, errors = await resolve_dependencies(dependant)
            handle_errors(errors)

            return await cast(Callable[..., Coroutine[Any, Any, T]], dependant.call)(*args, **{**deps, **kwargs})

        @wraps(func)
        def sync_call_with_solved_dependencies(*args: Any, **kwargs: Any) -> T:  # noqa: ANN401
            dependant = get_dependant(path="command", call=func)
            validate_dependant(dependant)
            deps, errors = asyncio.run(resolve_dependencies(dependant))
            handle_errors(errors)

            return cast(Callable[..., T], dependant.call)(*args, **{**deps, **kwargs})

        return async_call_with_solved_dependencies if is_async else sync_call_with_solved_dependencies

    if func is None:
        return _impl  # type: ignore  # noqa: PGH003
    return _impl(func)
```

I apologize for the use of many `noqa` or `type: ignore` to reduce boilerplate code; if anyone would like to improve this version, you’re welcome 😃

Next, let's do a small experiment:

```python
from typing import Annotated
from fastapi import Depends

class Brand:
    pass

class Car:
    def __init__(self, brand: Brand) -> None:
        self.brand = brand

def get_brand() -> Brand:
    return Brand()

def get_car(brand: Annotated[Brand, Depends(get_brand)]) -> Car:
    return Car(brand)
```

What happens when you directly call `get_brand` and `get_car`?

```python
brand = get_brand()  # No Error since this function doesn't have any dependencies
car = get_car()  # TypeError: get_car() missing 1 required positional argument: 'brand'
```

As expected, `get_car()` breaks since it requires a parameter, which brings us to why we’re here, right?

By using `injectable`, we can achieve two outcomes to meet your requirements:

## 1. Adding `@injectable` to the Function

### When `use_cache=True` (default behavior)

```python
@injectable  # Equals to @injectable(use_cache=True)
def get_car(brand: Annotated[Brand, Depends(get_brand)]) -> Car:
    return Car(brand)

car_1 = get_car()  # <__main__.Car at 0x10620cf20>
car_2 = get_car()  # <__main__.Car at 0x10953c140>
car_3 = get_car()  # <__main__.Car at 0x1095b7e00>
assert car_1.brand is car_2.brand is car_3.brand  # True
```

Notice that the `Car` instances share the same `brand` instance.

### When `use_cache=False`

```python
@injectable(use_cache=False)
def get_car(brand: Annotated[Brand, Depends(get_brand)]) -> Car:
    return Car(brand)


car_1 = get_car()  # <__main__.Car at 0x10620cf20>
car_2 = get_car()  # <__main__.Car at 0x10953c140>
car_3 = get_car()  # <__main__.Car at 0x1095b7e00>
assert car_1.brand is not car_2.brand is not car_3.brand  # True
```

Notice that each `Car` instance has a different `brand` instance.

## 2. Wrapping Your Function with `injectable`

This is especially useful when you cannot redefine the function.

### When `use_cache=True` (default behavior)

```python
def get_car(brand: Annotated[Brand, Depends(get_brand)]) -> Car:
    return Car(brand)

injectable_get_car = injectable(get_car)  # Equals to injectable(get_car, use_cache=True)
car_1 = injectable_get_car()  # <__main__.Car at 0x10620cf20>
car_2 = injectable_get_car()  # <__main__.Car at 0x10953c140>
car_3 = injectable_get_car()  # <__main__.Car at 0x1095b7e00>
assert car_1.brand is car_2.brand is car_3.brand  # True
```

### When `use_cache=False`

```python
def get_car(brand: Annotated[Brand, Depends(get_brand)]) -> Car:
    return Car(brand)

injectable_get_car = injectable(get_car, use_cache=False)
car_1 = injectable_get_car()  # <__main__.Car at 0x10620cf20>
car_2 = injectable_get_car()  # <__main__.Car at 0x10953c140>
car_3 = injectable_get_car()  # <__main__.Car at 0x1095b7e00>
assert car_1.brand is not car_2.brand is not car_3.brand  # True
```

---

So far, the only known issue is the lack of support for Static Code Analyzers, but for me, that’s acceptable. Again, I’d like to thank @barapa for the initial version, and I hope this helps others!

**Reply by JasperSui** [reply]
I made a blog post if you wanna get bonus test cases:
- `en` version: https://j-sui.com/2024/10/26/use-fastapi-depends-outside-fastapi-routes-en/
- `zh-TW` version: https://j-sui.com/2024/10/26/use-fastapi-depends-outside-fastapi-routes/

**Reply by edwinmarinn** [reply]
Great post, congrats! 

Just one question:
Can this decorator handle dependencies with yield?

References: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/

**Reply by JasperSui** [reply]
Thanks @edwinmarinn!

> Can this decorator handle dependencies with yield?

Yeah I think so.

I don't have the environment to test right now, I remember I had some dependencies with yield and they worked well without raising exceptions a.k.a. getting those dependecy instances successfully, but I am not sure code after `yield` will be executed correctly since it depends on the scenarios. I guess we should handle that manually before the process being gracefully killed by something like registering some callbacks on signals.

**Reply by abhiaagarwal** [reply]
Super useful, thanks! To get your type-checker to shut up, add this at the start:

```python
if TYPE_CHECKING:
    @overload
    def injectable(
        func: Callable[..., Coroutine[Any, Any, T]],
    ) -> Callable[..., Coroutine[Any, Any, T]]: ...

    @overload
    def injectable(func: Callable[..., T]) -> Callable[..., T]: ...
```

Note that order matters since the first overload is technically more specific than the second.

**Reply by JasperSui** [reply]
Glad to hear that @abhiaagarwal! I was trying to overload the function for the correct type hints but I couldn't get it work perfectly, I'll definitely try yours, thanks bro

**Reply by JasperSui** [reply]
@edwinmarinn I am facing this problem and here's some updates.

> Can this decorator handle dependencies with yield?

No, in fact, I think the decorator can't handle the depedencies with yield itself, because this is not its responsibility, instead, we need another helper function to gracefully execute the rest parts of code after `yield`, you may imagine some function like this:

```python
def get_session():
    session = Session() # Fake
    try:
        yield session
    except:
        session.rollback()
    finally:
        session.commit()
```

It's quite common to have functions like this, but with my previous implmentation, we can't control or execute the parts after `yield session`, and it can be dangerous, in this case, if the parts after `yield` are not executed, the changes made by this session will just be lost.

---
I've came up with a more robust solution to handle the parts after `yield` and I'll post it tomorrow, again, we can handle the parts, but not handle by the `injectable`, we will make use of the `AsyncExitStack` and modify a little bit of `injectable` to keep the call stack, then, we can have another function maybe like `teardown_injected` to iterate the call stack to close all generator one by one.

Once we have the function, we can register it to `atexit` and `signal`, or choose whenever we want to teardown the injected objects for graceful shutdown correctly, in my case, I might wanna `teardown_injected` once the worker finishes the queue message process, to behave like how FastAPI does when processing http requests.


---
**Updates 2024/12/17 04:32 (UTC+0)**
I'll make a dedicated package for the final version.

**Reply by JasperSui** [reply]
Hey guys, I just officially publish the [fastapi-injectable](https://github.com/JasperSui/fastapi-injectable) package.

Basic example here:
```python
from typing import Annotated

from fastapi import Depends
from fastapi_injectable import injectable

class Database:
    def query(self) -> str:
        return "data"

def get_db() -> Database:
    return Database()

@injectable
def process_data(db: Annotated[Database, Depends(get_db)]) -> str:
    return db.query()

# Use it anywhere!
result = process_data()
print(result) # Output: 'data'
```

Install with: `pip install fastapi-injectable`

The package lets you use FastAPI's dependency injection in:
- CLI tools
- Background workers
- Test fixtures
- Any non-HTTP context

I'd love to hear your feedback and contributions to make this package better! Feel free to open issues or PRs if you have ideas for improvements!

**Reply by edwinmarinn** [reply]
Awesome!
I'll give it a look

**Comment by agn-7** [comment]
I already have created a [post on StackOverflow](https://stackoverflow.com/questions/77800955/how-to-use-depends-or-similar-thing-as-dependency-injection-outside-of-fastapi-r) in this regard:

Here's my solution:

```python
class Provide:
    def __init__(self, value):
        self.value = value

def inject_db(f):
    sig = inspect.signature(f)
    async def wrapper(*args, **kwargs):
        for param in sig.parameters.values():
            if isinstance(param.default, Provide):
                async for db in param.default.value():
                    kwargs[param.name] = db
                    await f(*args, **kwargs)
    return wrapper

@cli.command(name="create_superuser")
@inject_db
async def create_superuser(db: database.AsyncSession = Provide(database.get_db)):
    username = click.prompt("Username", type=str)
    email = click.prompt("Email (optional)", type=str, default="")
    password = getpass("Password: ")
    confirm_password = getpass("Confirm Password: ")

    user = schemas.UserAdminCreate(
        username=username,
        email=None if not email else email,
        password=password,
        role="admin",
    )
```

**Comment by JasperSui** [comment]
Hey guys, I just officially publish the [fastapi-injectable](https://github.com/JasperSui/fastapi-injectable) package.

Basic example here:
```python
from typing import Annotated

from fastapi import Depends
from fastapi_injectable import injectable

class Database:
    def query(self) -> str:
        return "data"

def get_db() -> Database:
    return Database()

@injectable
def process_data(db: Annotated[Database, Depends(get_db)]) -> str:
    return db.query()

# Use it anywhere!
result = process_data()
print(result) # Output: 'data'
```

Install with: `pip install fastapi-injectable`

The package lets you use FastAPI's dependency injection in:
- CLI tools
- Background workers
- Test fixtures
- Any non-HTTP context

I'd love to hear your feedback and contributions to make this package better! Feel free to open issues or PRs if you have ideas for improvements!
